ofp4/ofvisitors.cpp

```diff
@@ -62,31 +64,74 @@ bool OpenFlowPrint::preorder(const IR::OF_Slice* e)  {
     return false;
 }
 
-bool OpenFlowPrint::preorder(const IR::OF_EqualsMatch* e)  {
-    auto reg = e->left->to<IR::OF_Register>();
-    if (reg != nullptr && reg->isSlice()) {
-        /* field=value/mask */
-        visit(e->left);
-        buffer += "=";
-        if (reg->low) {
-            buffer += "${";
-            if (e->right->to<IR::OF_Constant>()) {
-                visit(e->right);
-            } else if (auto value = e->right->to<IR::OF_InterpolatedVarExpression>()) {
+// 'erms' must have at least one element.  All its elements must have 'left'
+// that are disjoint slices of the same OF_Register.
+static void printRegisterMatch(std::vector<const IR::OF_EqualsMatch*>& erms,
+                               OpenFlowPrint& ofp,
+                               std::string& buffer)  {
+    auto reg0 = erms[0]->left->checkedTo<IR::OF_Register>();
+
+    /* field=value/mask */
+    if (erms.size() > 1 || reg0->friendlyName.isNullOrEmpty()) {
+        buffer += reg0->name;
+    } else {
+        buffer += "${r_" + reg0->friendlyName + "(true)}";
+    }
+    buffer += "=";
+    IR::Constant mask = 0;
+    if (erms.size() == 1 && !reg0->low) {
+        ofp.visit(erms[0]->right);
+        mask = reg0->mask();
+    } else {
+        buffer += "${";
+
+        size_t n = 0;
+        for (auto erm : erms) {
+            auto reg = erm->left->checkedTo<IR::OF_Register>();
+            if ((mask & reg->mask()).value != 0) {
+                BUG("%1%: overlapping bitwise matches on register", reg);
```

Comment:
I think this should be an error rather than a bug. People could easily write a program that requires overlapping masks. Perhaps in this case we should do something in the lowering pass.

Reply:
Isn't this indeed a bug in the compiler if it triggers? Maybe I don't understand the difference between BUG and `error`.

Reply:
BUG = assert
You don't call assert for inputs that cannot validate, only for internal logic errors.